### PR TITLE
Add device plugin doc under concepts/cluster-administration.

### PR DESCRIPTION
**Pull request by jiayingz**
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> For 1.8 Features: set Milestone to `1.8` and Base Branch to `release-1.8`
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>
> NOTE: Please check the “Allow edits from maintainers” box (see image below) to
> [allow reviewers to fix problems](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) on your patch and speed up the review process.
>
> Please delete this note before submitting the pull request.

Allow edits from maintainers checkbox

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/5261)

### CODE REVIEW

File: docs/concepts/cluster-administration/device-plugins.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    E.g., Nvidia GPUs are advertised as `nvidia.com/gpu`
+
+Following a successful registration, the device plugin will send Kubelet the
+list of devices it manages, and Kubelet will be in charge of advertising those
+resources to the API server as part of the Kubelet node status update.
+E.g., after a device plugin registers `vendor-domain/foo` with Kubelet
+and reports two healthy devices on a node, the node status will be updated
+to advertise 2 `vendor-domain/foo`.
+Devices can then be selected using the same process as for OIRs in the pod spec.
+Currently, extended resources are only uspported as integer resources and expect
```
Comment (tengqm) [inline docs/concepts/cluster-administration/device-plugins.md]: uspported -> supported

File: docs/concepts/cluster-administration/device-plugins.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+## Overview
+The 1.8 Kubernetes release supports device plugin as an alpha feature that
+is gated by DevicePlugins feature gate and is disabled by default.
+When the DevicePlugins feature is enabled, Kubelet will export a `Registration` gRPC service:
+```gRPC
+service Registration {
+	rpc Register(RegisterRequest) returns (Empty) {}
+}
+```
+A device plugin can register itself to Kubelet through this gRPC service
```
Comment (tengqm) [inline docs/concepts/cluster-administration/device-plugins.md]: add a period to this sentence

File: docs/concepts/cluster-administration/device-plugins.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+E.g., after a device plugin registers `vendor-domain/foo` with Kubelet
+and reports two healthy devices on a node, the node status will be updated
+to advertise 2 `vendor-domain/foo`.
+Devices can then be selected using the same process as for OIRs in the pod spec.
+Currently, extended resources are only uspported as integer resources and expect
+to always have limits == requests in container resource Spec.
+
+## Device Plugin Implementation
+
+The general workflow of a device plugin includes the following parts:
```
Comment (tengqm) [inline docs/concepts/cluster-administration/device-plugins.md]: parts -> steps/phases?

File: docs/concepts/cluster-administration/device-plugins.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+* Initialization. During this phase, a device plugin performs vendor specific initialization and setup to make sure their devices are in ready state.
```
Comment (tengqm) [inline docs/concepts/cluster-administration/device-plugins.md]: a device plugin -> the device plugin

their devices -> the devices

File: docs/concepts/cluster-administration/device-plugins.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+* Starts a gRPC service with a unix socket under host path: `/var/lib/kubelet/device-plugins/` that implements the following interfaces:
```
Comment (tengqm) [inline docs/concepts/cluster-administration/device-plugins.md]: The plugin starts a ...

File: docs/concepts/cluster-administration/device-plugins.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        // returns the new list
+        rpc ListAndWatch(Empty) returns (stream ListAndWatchResponse) {}
+
+        // Allocate is called during container creation so that the Device
+        // Plugin can run device specific operations and instruct Kubelet
+        // of the steps to make the Device available in the container
+        rpc Allocate(AllocateRequest) returns (AllocateResponse) {}
+}
+```
+* Register itself with Kubelet through unix socket at host path `/var/lib/kubelet/device-plugins/kubelet.sock`.
```
Comment (tengqm) [inline docs/concepts/cluster-administration/device-plugins.md]: The plugin registers itself ...

File: docs/concepts/cluster-administration/device-plugins.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+* After successfully registering itself, the device plugin runs in serving mode during which it keeps
+monitoring device health and reports back to Kubelet upon any device state changes.
+It is also responsible for serving Allocate gRPC request. During Allocate, the device plugin may
```
Comment (tengqm) [inline docs/concepts/cluster-administration/device-plugins.md]: request -> requests

File: docs/concepts/cluster-administration/device-plugins.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+perform device specific preparing operations (e.g., gpu cleanup, QRNG initialization, and etc.).
+If the operations succeed, the device plugin will return AllocateResponse that contains container
```
Comment (tengqm) [inline docs/concepts/cluster-administration/device-plugins.md]: return -> return an

File: docs/concepts/cluster-administration/device-plugins.md (hunk 2 above)
Comment (jiayingz) [reply docs/concepts/cluster-administration/device-plugins.md]: done.

File: docs/concepts/cluster-administration/device-plugins.md (hunk 1 above)
Comment (jiayingz) [reply docs/concepts/cluster-administration/device-plugins.md]: done.

File: docs/concepts/cluster-administration/device-plugins.md (hunk 4 above)
Comment (jiayingz) [reply docs/concepts/cluster-administration/device-plugins.md]: done.

File: docs/concepts/cluster-administration/device-plugins.md (hunk 3 above)
Comment (jiayingz) [reply docs/concepts/cluster-administration/device-plugins.md]: done.

File: docs/concepts/cluster-administration/device-plugins.md (hunk 5 above)
Comment (jiayingz) [reply docs/concepts/cluster-administration/device-plugins.md]: done.

File: docs/concepts/cluster-administration/device-plugins.md (hunk 7 above)
Comment (jiayingz) [reply docs/concepts/cluster-administration/device-plugins.md]: done.

File: docs/concepts/cluster-administration/device-plugins.md (hunk 8 above)
Comment (jiayingz) [reply docs/concepts/cluster-administration/device-plugins.md]: done.

**Comment by mindprince** [comment]
Who is the audience of this document? Device plugin vendors? Cluster administrators? Users?

**Comment by mindprince** [comment]
Code is not rendered correctly: https://deploy-preview-5261--kubernetes-io-vnext-staging.netlify.com/docs/concepts/cluster-administration/device-plugins/

**Comment by steveperry-53** [comment]
@jiayingz, It would be nice if I could do an edit pass on this topic. Is the **Allow edits by maintainers** checkbox available in the right column? If so, could you check it. Thanks.

**Comment by steveperry-53** [comment]
@jiayingz, See https://github.com/kubernetes/kubernetes.github.io/pull/5261/commits/f509e166f8ddd380c182ff42049b9368d609bc35 for suggested edits.

Next I need to adjust the Alpha notice to say 1.8 instead of 1.7.

**Comment by steveperry-53** [comment]
@jiayingz, Is DaemonSet vs. Bare Metal the correct distinction? Here's how I see it: Regardless of whether you're on bare metal or a VM, you can deploy your device plugin as a DaemonSet (or not). Maybe we should say something like, "A device plugin can run as a DaemonSet or as a manually deployed component."

**Comment by jiayingz** [comment]
I am fine with changing "bare metal" to "as a manually deployed component". Agree it is more clear.

**Comment by steveperry-53** [comment]
@jiayingz, See https://github.com/kubernetes/kubernetes.github.io/pull/5261/commits/cedc291ef86f1319acb07ddd62c64be1864ca780 and https://github.com/kubernetes/kubernetes.github.io/pull/5261/commits/1561b1751104c7ca60b5602bc8a7ed5ae8878204 for a few more edits.
After https://github.com/kubernetes/kubernetes.github.io/pull/5341 is merged, the alpha notice should change from 1.7 to 1.8.
Let me know when you're ready for me to merge this.

**Comment by jiayingz** [comment]
lgtm

**Pull request merged**